mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp

```diff
@@ -1716,29 +1735,35 @@ struct ReduceConstantOptimization : public OpRewritePattern<OperationType> {
     const auto shapedOldElementsValues =
         denseElementsAttr.getType().cast<ShapedType>();
 
-    if (!llvm::isa<IntegerType>(shapedOldElementsValues.getElementType()))
-      return rewriter.notifyMatchFailure(
-          op, "reduce input currently supported with integer type");
+    if constexpr (std::is_same_v<OperationType, ReduceAllOp> ||
+                  std::is_same_v<OperationType, ReduceAnyOp>) {
+      if (!llvm::isa<IntegerType>(shapedOldElementsValues.getElementType()))
+        return rewriter.notifyMatchFailure(
+            op, "reduce input currently supported with integer type");
+    }
 
     auto oldShape = shapedOldElementsValues.getShape();
     auto newShape = resultType.getShape();
 
     auto newNumOfElements = std::accumulate(newShape.begin(), newShape.end(), 1,
                                             std::multiplies<int>());
-    llvm::SmallVector<APInt> newReducedTensor(newNumOfElements);
-
-    for (int64_t reductionIndex = 0; reductionIndex < newNumOfElements;
-         ++reductionIndex) {
+    auto rankedTensorType = cast<RankedTensorType>(resultType);
 
-      /// Let's reduce all the elements along this reduction axis
-      newReducedTensor[reductionIndex] = calculateReducedValue<OperationType>(
-          denseElementsAttr, oldShape, reductionAxis, reductionIndex);
+    DenseElementsAttr resultDenseAttr;
+    if (llvm::isa<IntegerType>(shapedOldElementsValues.getElementType())) {
+      llvm::SmallVector<APInt> newReducedTensor(newNumOfElements);
+      resultDenseAttr =
+          this->compute<APInt>(newReducedTensor, denseElementsAttr, oldShape,
+                               reductionAxis, rankedTensorType);
+    } else if (llvm::isa<FloatType>(shapedOldElementsValues.getElementType())) {
+      llvm::SmallVector<APFloat> newReducedTensor(newNumOfElements,
+                                                  APFloat(0.0));
+      resultDenseAttr =
+          this->compute<APFloat>(newReducedTensor, denseElementsAttr, oldShape,
+                                 reductionAxis, rankedTensorType);
```

Comment:
I don't know if it will happen of element type not be Integer or Float, but can we have an else statement here just to be safe? something like:
```suggestion
                                 reductionAxis, rankedTensorType);
                          } else {
                        return rewriter.notifyMatchFailure("unsupported element type");
                          
                          }
```